### PR TITLE
fix: advance async read when msg too big

### PR DESCRIPTION
**Pull request by niklasad1**
Closing #53

### CODE REVIEW

**Review by dvdplm — APPROVED** [review]

**Review by jsdw — APPROVED** [review]
The code looks good to me!

Is there a test somewhere to make sure that this is working? 

Now I can see the context, I'm wondering about the performance implications; I can imagine that if a message is say 1MB over, it'll take 1000 loop iterations and may be quite slow. Maybe we don't care too much?

One way to potentially give it a nice performance boost would be to create a `DiscardBytes` struct and impl `AsyncRead` for it so that `poll_read` does something similar to the approach here internally, but keeps looping until `Poll::Pending` is returned. This way, we'd discard as many bytes as we can in one "await", and avoid yielding to the futures runtime as much.

**Comment by niklasad1** [comment]
> Now I can see the context, I'm wondering about the performance implications; I can imagine that if a message is say 1MB over, it'll take 1000 loop iterations and may be quite slow. Maybe we don't care too much?

I benched against it `self.reader.read_exact(&mut vec![0; length]).await?;` and I couldn't find any significant difference in throughput it was similar for 1MB, 10MB and 100MB. It was around +/- 10%, in some scenarios array faster and the otherway around.

At least it's faster than reconnect to the soketto server,  as the soketto server will be in a bad state :P

I just tested it from `jsonrpsee` but there is not that many tests in this repo, I assumed we rely on the autobahn test suite...

**Comment by niklasad1** [comment]
> One way to potentially give it a nice performance boost would be to create a DiscardBytes struct and impl AsyncRead for it so that poll_read does something similar to the approach here internally, but keeps looping until Poll::Pending is returned. This way, we'd discard as many bytes as we can in one "await", and avoid yielding to the futures runtime as much.

Good idea.

**Comment by niklasad1** [comment]
@jsdw I managed to use the take trick instead and it was indeed 10-15% faster

File: src/connection.rs [hunk 1]
```diff
@@ -686,3 +689,24 @@ impl From<base::Error> for Error {
 		Error::Codec(e)
 	}
 }
+
+/// Discard `n` bytes from underlying reader.
```
Comment (niklasad1) [inline src/connection.rs]:
```suggestion
/// Discard `n` bytes from the underlying reader.
```

**Review by dvdplm — APPROVED** [review]
Neat!

**Pull request merged**